Adding new version update scripts

garrettjonesgoogle opened the pull request:
versions.txt has both the last-released version (in maven) and the current version (either about-to-be-released or the snapshot version). 

There are two scripts:

- bump_versions.py: Bumps the versions stored in version.txt
- replace_versions.py: Idempotent script which replaces the versions everywhere they are tagged

Users can choose to manually update versions.txt if they so choose and skip the bump_versions.py script.

vam-google commented on utilities/bump_versions.py:
```diff
@@ -0,0 +1,111 @@
+        self.snapshot = snapshot
+
+    def __str__(self):
+        mmp = str(self.major) + '.' + str(self.minor) + '.' + str(self.patch)
```
please use 
```python
"%s%s" % (var1, var2)
```
for string concatenation. This applies to all other places where `+` is used.

garrettjonesgoogle replied:
I'll use the newer idiom `'{}'.format(x)`

vam-google commented on utilities/bump_versions.py:
```diff
@@ -0,0 +1,111 @@
+# Copyright 2018 Google LLC
```
This version of python is this supported for? Please make sure that it runs well on 2.7 and 3.4+ normally.

garrettjonesgoogle replied:
I just now tested on python 2.x and 3.x and it seems to work fine.

vam-google commented on utilities/bump_versions.py:
```diff
@@ -0,0 +1,111 @@
+import argparse
+import copy
+import re
+
```
Please add documentation about how to use this script with examples of typical usage

garrettjonesgoogle replied:
Adding.

vam-google commented on utilities/bump_versions.py:
```diff
@@ -0,0 +1,111 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import argparse
```
I'm still not convinced that we need the bumping script at all. Find/raplace is easy in the version.txt file.

garrettjonesgoogle replied:
If we don't have a bumping script, then we need to write documentation that explains the right process for bumping versions (i.e. turn this script into human instructions). This could be a pretty error-prone process for newbies; I want the instructions for standard releases to be easy.

vam-google commented on utilities/replace_versions.py:
```diff
@@ -0,0 +1,108 @@
+# Copyright 2018 Google LLC
```
Please add documentation about how to use this script with examples of typical usage.

vam-google commented on utilities/replace_versions.py:
```diff
@@ -0,0 +1,108 @@
+version_regex_str = r'\d+\.\d+\.\d+(?:-\w+)?(?:-\w+)?'
+version_regex = re.compile(version_regex_str)
+
+class CodeModule:
```
This looks like java way of doing it (a class without methods). In Python/dynamic languages you usually just have a dictionary. yes, there will be static structure, but python is not compilable/static language so there is no expectations of having a defined structure.

garrettjonesgoogle replied:
1. This is google-cloud-java. We are java people :-P
2. It cleans up other code by taking care of processing and cleaning up the input

vam-google replied:
Python people are in a towering rage...

garrettjonesgoogle replied:
¯\_(ツ)_/¯

They can stick to google-cloud-python :-)

vam-google commented on utilities/replace_versions.py:
```diff
@@ -0,0 +1,108 @@
+    if target:
+        replace_versions(version_map, target)
+    else:
+        for root, dirs, files in os.walk("."):
```
Please mention in the doc that the file must be executed from repo root like: 
```
> python utilities/replace_versions.py
```
otherwise this would not work

garrettjonesgoogle replied:
Will do.

vam-google commented on utilities/replace_versions.py:
```diff
@@ -0,0 +1,108 @@
+                else:
+                    match = version_update_end_marker.search(line)
+                    if match:
+                        repl_open = False
```
In python you can do it like this (and I think community preferes this way):
```python
repl_open, repl_thisline = False, False
```

This applies to all similar blocks of initialization, and allows making the script substantially shorter.

garrettjonesgoogle replied:
I'll make this change

garrettjonesgoogle commented:
I have addressed all feedback and tagged all versions. I verified it is correct by running bump_versions.py twice and grepping for the old version to make sure it appears nowhere. I also decided to solve the nio-examples case by just having it use "current", which means it will be set to the correct thing after a release when the releasing person calls `bump_version.py next_snapshot patch`. PTAL

vam-google commented on google-cloud-contrib/google-cloud-nio-examples/pom.xml:
```diff
@@ -12,7 +12,7 @@
     <groupId>com.google.cloud</groupId>
     <artifactId>google-cloud-contrib</artifactId>
-    <version>0.35.1-alpha-SNAPSHOT</version>
+    <version>0.35.1-alpha-SNAPSHOT</version><!-- {x-version-update:google-cloud-pom:current} -->
```
Should it be `google-cloud-contrib:current`?

garrettjonesgoogle replied:
good catch, will fix

vam-google commented on google-cloud-contrib/google-cloud-logging-logback/pom.xml:
```diff
@@ -20,7 +20,7 @@
     <groupId>com.google.cloud</groupId>
     <artifactId>google-cloud-contrib</artifactId>
-    <version>0.35.1-alpha-SNAPSHOT</version>
+    <version>0.35.1-alpha-SNAPSHOT</version><!-- {x-version-update:google-cloud-pom:current} -->
```
Should it be `google-cloud-contrib:current`?

vam-google commented on google-cloud-contrib/google-cloud-nio-examples/README.md:
```diff
@@ -21,6 +21,7 @@ To run this example:
 
 4.    Run the sample with:
 
+[//]: # ({x-version-update-start:google-cloud-nio:current})
```
Should this be `google-cloud-nio` or `google-cloud-nio-examples`?

garrettjonesgoogle replied:
Hmm this one is tricky - there is a single line which contains both versions. I'm not sure how to address this intra-line difference, besides just require that nio & nio-examples match versions.

vam-google commented on utilities/update_versions.sh:
```diff
@@ -1,278 +0,0 @@
-#!/bin/bash
```
My masterpiece is dying... It was ahead of its time, too masterpiecy for its contemporaries.... =(...

garrettjonesgoogle replied:
Naw, it's just that you chose bash as the language of your masterpiece :-P

garrettjonesgoogle commented:
versions.txt has both the last-released version (in maven) and the current version (either about-to-be-released or the snapshot version).

There are two scripts:

bump_versions.py: Bumps the versions stored in version.txt
replace_versions.py: Idempotent script which replaces the versions everywhere they are tagged
Users can choose to manually update versions.txt if they so choose and skip the bump_versions.py script.
